jormungandr/src/client.rs

```diff
@@ -105,22 +106,59 @@ async fn get_block_tip(blockchain_tip: Tip) -> Header {
     tip.header().clone()
 }
 
-async fn handle_get_headers_range(
+fn get_block_from_storage(storage: &Storage, id: HeaderHash) -> Result<Block, Error> {
+    match storage.get(id) {
+        Ok(Some(block)) => Ok(block),
+        Ok(None) => Err(Error::not_found(format!(
+            "block {} is not known to this node",
+            id
+        ))),
+        Err(e) => Err(e.into()),
+    }
+}
+
+async fn fuse_send_items<T, V>(
+    items: T,
+    reply_handle: ReplyStreamHandle<V>,
+) -> Result<(), ReplySendError>
+where
+    T: IntoIterator<Item = Result<V, Error>>,
+{
+    let mut sink = reply_handle.start_sending();
+    for item in items.into_iter() {
+        let err = item.is_err();
+        sink.feed(item).await?;
+        if err {
+            break;
+        }
+    }
+    sink.close().await
+}
+
+async fn send_range_from_storage<T, F>(
     storage: Storage,
-    checkpoints: Vec<HeaderHash>,
+    from: Vec<HeaderHash>,
```

Comment:
what is the difference? `PullHeaders`, `PullBlocks` and `PullBlocksToTip` have all the same behavior but that field is called differently in some of them. 

On that subject, why is `PullHeaders`(gRPC proto name) renamed in `GetHeadersRange` (`ClientMsg` variant)? I would like to uniform those as well if possible

Reply:
> what is the difference?

That is just more self-descriptive. The logic behind this is that the node is checking all ids in the list and starts from the first id present on this node. So they are exactly "checkpoints".

As per `GetHeadersRange`: because this is what it does. It is getting some range of headers. Now it should be used everywhere and `PullBlocksToTip` should not be used anymore and be removed entirely later.

Reply:
So this rename was basically done to avoid confusion with old implementations.

Reply:
Should we rename `PullBlocks`  to `GetBlocksRange`, and all `from` parameters to `checkpoints` as well?

Reply:
It's OK to change the semantics of the existing methods, we are in pre-0.10 phase as far as the world is concerned and we can break the API. I kinda like `PullBlocks` because I authored the API, the name's shorter, and renaming would have to be done everywhere because it percolates into the generated code and we wouldn't want to have differently named methods in the protocol abstraction mini-layer.

Names of the fields do not matter as much, they are numbers in protobuf and we expose them as function parameters in the application-facing API. I defer to @eugene-babichenko on the non-confusingness of `checkpoints` :)

Reply:
> I kinda like PullBlocks because I authored the API, the name's shorter, and renaming would have to be done everywhere because it percolates into the generated code and we wouldn't want to have differently named methods in the protocol abstraction mini-layer.

That why I find kind of confusing the fact that `PullHeaders` and `GetHeadersRange` refer to the same action but have different names. 
I would like to use consistent naming and have no real preference, but as Mikhail noted changing the gRPC name is probably more difficult.

Reply:
Well, yeah, re-checked the thread and the code, agree on `PullHeaders`. Also, there is still a typo to fix :)
As per `from` vs `checkpoints` - we may leave it as is (that is why I approved the PR, doesn't matter that much) for now for the sake of consistency. I re-checked we are really using `from` almost everywhere. The term `checkpoint` is just kind of appealing in this context, because it better reflects the nature of the variable. But renaming everything afterwards for consistency is bothersome, so let's leave it this way.